clients/rospy/src/rospy/impl/tcpros_pubsub.py

```diff
@@ -386,15 +386,17 @@ def __init__(self, connection, queue_size):
 
         self._lock = threading.Lock()
         self._cond_data_available = threading.Condition(self._lock)
-        self._connection.set_cleanup_callback(self._closed_connection_callback)
         self._queue = []
         self._error = None
 
         self._thread = threading.Thread(target=self._run)
         self._thread.start()
 
+        self._connection.add_cleanup_callback(self._closed_connection_callback)
+
     def _closed_connection_callback(self, connection):
-        self._cond_data_available.notify()
+        with self._lock:
```

Comment:
It silently failed without it.  The Lock must be acquired before you can access the Condition.  See https://docs.python.org/2/library/threading.html#condition-objects

"If the calling thread has not acquired the lock when this method is called, a RuntimeError is raised."